net/instaweb/http/redirect_following_url_async_fetcher.cc

```diff
@@ -0,0 +1,294 @@
+#include "net/instaweb/rewriter/public/rewrite_options.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/null_message_handler.h"
+#include "pagespeed/kernel/base/string.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/kernel/http/google_url.h"
+
+// TODO(oschaaf): inlining & intent should  be persisted across redirects.
+namespace net_instaweb {
+
+class RedirectFollowingUrlAsyncFetcher::RedirectFollowingFetch
+    : public SharedAsyncFetch {
+ public:
+  RedirectFollowingFetch(
+      RedirectFollowingUrlAsyncFetcher* redirect_following_fetcher,
+      AsyncFetch* base_fetch, const GoogleString& url,
+      const GoogleString& context_url, MessageHandler* message_handler)
+      : SharedAsyncFetch(base_fetch),
+        redirect_following_fetcher_(redirect_following_fetcher),
+        recieved_redirect_status_code_(false),
+        urls_seen_(new GoogleUrlSet()),
+        url_(url),
+        gurl_(url),
+        base_fetch_(base_fetch),
+        context_url_(context_url),
+        message_handler_(message_handler) {
+    GoogleString sanitized = GoogleUrl::Sanitize(url);
+    urls_seen_->insert(sanitized);
+  }
+
+  RedirectFollowingFetch(
+      RedirectFollowingUrlAsyncFetcher* redirect_following_fetcher,
+      AsyncFetch* base_fetch, const GoogleString& url,
+      const GoogleString& context_url, GoogleUrlSet* redirects_followed_earlier,
+      MessageHandler* message_handler)
+      : SharedAsyncFetch(base_fetch),
+        redirect_following_fetcher_(redirect_following_fetcher),
+        recieved_redirect_status_code_(false),
+        urls_seen_(redirects_followed_earlier),
+        url_(url),
+        gurl_(url),
+        base_fetch_(base_fetch),
+        context_url_(context_url),
+        message_handler_(message_handler) {}
+
+  bool Validate() {
+    if (!gurl_.IsWebValid()) {
+      response_headers()->set_status_code(HttpStatus::kBadRequest);
+      return false;
+    }
+    return true;
+  }
+
+ protected:
+  virtual bool HandleFlush(MessageHandler* message_handler) {
+    if (!recieved_redirect_status_code_) {
+      return SharedAsyncFetch::HandleFlush(message_handler);
+    }
+    return true;
+  }
+
+  virtual void HandleHeadersComplete() {
+    GoogleString h = response_headers()->ToString();
+    // Currently we support permanent and temporary redirects.
+    recieved_redirect_status_code_ =
+        response_headers()->status_code() == HttpStatus::kMovedPermanently ||
+        response_headers()->status_code() == HttpStatus::kFound;
+
+    if (!recieved_redirect_status_code_) {
+      SharedAsyncFetch::HandleHeadersComplete();
+    }
+  }
+
+  virtual bool HandleWrite(const StringPiece& content,
+                           MessageHandler* handler) {
+    if (!recieved_redirect_status_code_) {
+      return SharedAsyncFetch::HandleWrite(content, handler);
+    }
+    return true;
+  }
+
+  virtual void HandleDone(bool success) {
```

Comment:
I feel like this function would benefit from breaking up into smaller methods for readability.